In [12]:
import os
import numpy as np
from PIL import Image
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

In [13]:
dataset_path = "./bisindo_dataset/"
# i've actually used a second dataset which is: https://www.kaggle.com/datasets/riestiyazain/bisindo2

In [14]:
# this will pixel per value as the feature  
def image_to_array(path):
    features = []
    size = (224,224)

    # # for testing purpose (1 image)
    # image = Image.open(path)
    # image = image.resize(size)
    # features.append(np.array(image).flatten())

    for x in path:
        image = Image.open(x)
        image = image.resize(size)
        features.append(np.array(image).flatten())

    return features

In [15]:
from tensorflow.keras.applications import Xception
from tensorflow.keras.applications.xception import preprocess_input
from tensorflow.keras.preprocessing import image

def deep_feature(path):
    model = Xception(weights="imagenet", include_top=False)
    features = []
    for x in path:
        img = image.load_img(x, target_size=(224, 224))
        x = image.img_to_array(img)
        x = np.expand_dims(x, axis=0)
        x = preprocess_input(x)
        features.append(model.predict(x).flatten())

    return features


In [16]:
labels = [] # labels is used as Y
img_dir = [] # contain all image directory

# this will take image dir and its label
for x in os.listdir(dataset_path):
    subpath = os.path.join(dataset_path, x)
    for y in os.listdir(subpath):
        labels.append(x)
        img_dir.append(os.path.join(subpath, y))

In [17]:
# features = image_to_array(img_dir)

In [18]:
feature2 = deep_feature(img_dir)

1/1 [==============================] - 0s 124ms/step


In [19]:
def training(features, labels, k):
    X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=0.2, random_state=42)
    
    # Feature normalization
    scaler = StandardScaler()
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.transform(X_test)
    
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)   

    # Predict image
    y_pred = knn.predict(X_test)

    # Evaluate model performance
    accuracy = accuracy_score(y_test, y_pred)
    print(f"- K:{k} | Accuracy: {accuracy}")


In [20]:
# training(features, 5)

In [21]:
# training(features, labels, 1)
# training(features, labels, 3)
# training(features, labels, 5)
# training(features, labels, 7)
# training(features, labels, 9)
# training(features, labels, 11)
# training(features, labels, 13)

In [22]:
training(feature2, labels, 1)
training(feature2, labels, 3)
training(feature2, labels, 5)
training(feature2, labels, 7)
training(feature2, labels, 9)
training(feature2, labels, 11)
training(feature2, labels, 13)

- K:1 | Accuracy: 0.9342105263157895
- K:3 | Accuracy: 0.9219600725952813
- K:5 | Accuracy: 0.911070780399274
- K:7 | Accuracy: 0.9137931034482759
- K:9 | Accuracy: 0.8992740471869328
- K:11 | Accuracy: 0.8915607985480943
- K:13 | Accuracy: 0.8833938294010889


testing1: bisindo_dataset1 no normalisation

Deep feature 
- K:1 | Accuracy: 0.8667642752562226
- K:3 | Accuracy: 0.8433382137628112
- K:5 | Accuracy: 0.8462664714494875
- K:7 | Accuracy: 0.8418740849194729
- K:9 | Accuracy: 0.8418740849194729
- K:11 | Accuracy: 0.8477306002928258
- K:13 |Accuracy: 0.8374816983894583

Raw feature
- K:1 | Accuracy: 0.3938506588579795
- K:3 | Accuracy: 0.32210834553440704
- K:5 | Accuracy: 0.2869692532942899
- K:7 | Accuracy: 0.28111273792093705
- K:9 | Accuracy: 0.26207906295754024
- K:11 | Accuracy: 0.24011713030746706
- K:13 | Accuracy: 0.22693997071742314


Testing2: bisindo_dataset1 normalisation

Deep feature

Raw feature
- K:1 | Accuracy: 0.3718887262079063
- K:3 | Accuracy: 0.2986822840409956
- K:5 | Accuracy: 0.2679355783308931
- K:7 | Accuracy: 0.26061493411420206
- K:9 | Accuracy: 0.26061493411420206
- K:11 | Accuracy: 0.24011713030746706
- K:13 | Accuracy: 0.20937042459736457

Testing3: bisindo_dataset2 no normalisation

Deep feature
- K:1 | Accuracy: 0.9342105263157895
- K:3 | Accuracy: 0.9219600725952813
- K:5 | Accuracy: 0.911070780399274
- K:7 | Accuracy: 0.9137931034482759
- K:9 | Accuracy: 0.8992740471869328
- K:11 | Accuracy: 0.8915607985480943
- K:13 | Accuracy: 0.8833938294010889

Raw feature
- K:1 | Accuracy: 0.573502722323049
- K:3 | Accuracy: 0.4895644283121597
- K:5 | Accuracy: 0.46960072595281305
- K:7 | Accuracy: 0.45054446460980035
- K:9 | Accuracy: 0.43466424682395643
- K:11 | Accuracy: 0.411070780399274
- K:13 | Accuracy: 0.4069872958257713